### PR TITLE
configurable polling interval

### PR DESCRIPTION
**Pull request by fananchong**

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -134,6 +135,9 @@ func StartWithConfig(loadAppConfig func() (*config.AppConfig, error)) (Client, e
 	configComponent := &notify.ConfigComponent{}
 	configComponent.SetAppConfig(c.getAppConfig)
 	configComponent.SetCache(c.cache)
+	if beforeAsyncLongPoll != nil {
```
Comment (zouyx) [inline client.go:138]: 为什么要加个方法？用 loadappconfig 不行吗？

Comment (fananchong) [reply client.go:138]: 以下是之前捕获的 race 堆栈：

```go
2022-05-10 20:28:15.
==================
WARNING: DATA RACE
Read at 0x00c0001e6838 by goroutine 52:
container/list.(*List).Len()
/usr/local/go/src/container/list/list.go:66 +0x73
github.com/apolloconfig/agollo/v4/storage.(*Cache).pushChange()
/data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/fananchong/agollo/v4@v4.1.2-0.20220504001018-a7beadea6359/storage/repository.go:418 +0x2c
github.com/apolloconfig/agollo/v4/storage.(*Cache).pushNewestChanges()
/data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/fananchong/agollo/v4@v4.1.2-0.20220504001018-a7beadea6359/storage/repository.go:411 +0x138
github.com/apolloconfig/agollo/v4/storage.(*Cache).UpdateApolloConfig()
/data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/fananchong/agollo/v4@v4.1.2-0.20220504001018-a7beadea6359/storage/repository.go:263 +0x224
github.com/apolloconfig/agollo/v4/component/notify.(*ConfigComponent).Start()
/data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/fananchong/agollo/v4@v4.1.2-0.20220504001018-a7beadea6359/component/notify/componet_notify.go:53 +0x192
github.com/apolloconfig/agollo/v4/component.StartRefreshConfig()
/data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/fananchong/agollo/v4@v4.1.2-0.20220504001018-a7beadea6359/component/common.go:27 +0x37
github.com/apolloconfig/agollo/v4.StartWithConfig·dwrap·1()
/data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/fananchong/agollo/v4@v4.1.2-0.20220504001018-a7beadea6359/client.go:137 +0x47

Previous write at 0x00c0001e6838 by main goroutine:
container/list.(*List).insert()
/usr/local/go/src/container/list/list.go:98 +0x33c
container/list.(*List).insertValue()
/usr/local/go/src/container/list/list.go:104 +0x138
container/list.(*List).PushBack()
/usr/local/go/src/container/list/list.go:155 +0x11b
github.com/apolloconfig/agollo/v4/storage.(*Cache).AddChangeListener()
/data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/fananchong/agollo/v4@v4.1.2-0.20220504001018-a7beadea6359/storage/repository.go:377 +0x1d
github.com/apolloconfig/agollo/v4.(*internalClient).AddChangeListener()
/data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/fananchong/agollo/v4@v4.1.2-0.20220504001018-a7beadea6359/client.go:246 +0x54
haidao/backend/core/configv2.(*cfgApollo).startCommonNamespace()
/tmp/gocbuildbe8172cde925haidao_backend_cross_season/core/configv2/cfg_apollo.go:104 +0x2b8
haidao/backend/core/configv2.(*cfgApollo).Init()
/tmp/gocbuildbe8172cde925haidao_backend_cross_season/core/configv2/cfg_apollo.go:59 +0x828
haidao/backend/core/configv2.Init()
/tmp/gocbuildbe8172cde925haidao_backend_cross_season/core/configv2/cfg.go:44 +0x4f7
haidao/backend/core/app.Run.func2()
/tmp/gocbuildbe8172cde925haidao_backend_cross_season/core/app/app.go:128 +0x8ec
github.com/urfave/cli/v2.(*App).RunContext()
/data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/urfave/cli/v2@v2.5.1/app.go:322 +0xdd5
github.com/urfave/cli/v2.(*App).Run()
/data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/urfave/cli/v2@v2.5.1/app.go:224 +0x16a8
haidao/backend/core/app.Run()
/tmp/gocbuildbe8172cde925haidao_backend_cross_season/core/app/app.go:167 +0x162c
main.main()
/tmp/gocbuildbe8172cde925haidao_backend_cross_season/cross_season/main.go:30 +0x50

Goroutine 52 (running) created at:
github.com/apolloconfig/agollo/v4.StartWithConfig()
/data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/fananchong/agollo/v4@v4.1.2-0.20220504001018-a7beadea6359/client.go:137 +0x696
haidao/backend/core/configv2.(*cfgApollo).startCommonNamespace()
/tmp/gocbuildbe8172cde925haidao_backend_cross_season/core/configv2/cfg_apollo.go:88 +0x14b
haidao/backend/core/configv2.(*cfgApollo).Init()
/tmp/gocbuildbe8172cde925haidao_backend_cross_season/core/configv2/cfg_apollo.go:59 +0x828
haidao/backend/core/configv2.Init()
/tmp/gocbuildbe8172cde925haidao_backend_cross_season/core/configv2/cfg.go:44 +0x4f7
haidao/backend/core/app.Run.func2()
/tmp/gocbuildbe8172cde925haidao_backend_cross_season/core/app/app.go:128 +0x8ec
github.com/urfave/cli/v2.(*App).RunContext()
/data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/urfave/cli/v2@v2.5.1/app.go:322 +0xdd5
github.com/urfave/cli/v2.(*App).Run()
/data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/urfave/cli/v2@v2.5.1/app.go:224 +0x16a8
haidao/backend/core/app.Run()
/tmp/gocbuildbe8172cde925haidao_backend_cross_season/core/app/app.go:167 +0x162c
main.main()
/tmp/gocbuildbe8172cde925haidao_backend_cross_season/cross_season/main.go:30 +0x50
```

storage.(*Cache).AddChangeListener() 与 storage.(*Cache).pushChange() 间有 race

Comment (zouyx) [reply client.go:138]: > 以下是之前捕获的 race 堆栈：
> 
> ```go
> 2022-05-10 20:28:15.
> ==================
> WARNING: DATA RACE
> Read at 0x00c0001e6838 by goroutine 52:
> container/list.(*List).Len()
> /usr/local/go/src/container/list/list.go:66 +0x73
> github.com/apolloconfig/agollo/v4/storage.(*Cache).pushChange()
> /data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/fananchong/agollo/v4@v4.1.2-0.20220504001018-a7beadea6359/storage/repository.go:418 +0x2c
> github.com/apolloconfig/agollo/v4/storage.(*Cache).pushNewestChanges()
> /data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/fananchong/agollo/v4@v4.1.2-0.20220504001018-a7beadea6359/storage/repository.go:411 +0x138
> github.com/apolloconfig/agollo/v4/storage.(*Cache).UpdateApolloConfig()
> /data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/fananchong/agollo/v4@v4.1.2-0.20220504001018-a7beadea6359/storage/repository.go:263 +0x224
> github.com/apolloconfig/agollo/v4/component/notify.(*ConfigComponent).Start()
> /data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/fananchong/agollo/v4@v4.1.2-0.20220504001018-a7beadea6359/component/notify/componet_notify.go:53 +0x192
> github.com/apolloconfig/agollo/v4/component.StartRefreshConfig()
> /data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/fananchong/agollo/v4@v4.1.2-0.20220504001018-a7beadea6359/component/common.go:27 +0x37
> github.com/apolloconfig/agollo/v4.StartWithConfig·dwrap·1()
> /data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/fananchong/agollo/v4@v4.1.2-0.20220504001018-a7beadea6359/client.go:137 +0x47
> 
> Previous write at 0x00c0001e6838 by main goroutine:
> container/list.(*List).insert()
> /usr/local/go/src/container/list/list.go:98 +0x33c
> container/list.(*List).insertValue()
> /usr/local/go/src/container/list/list.go:104 +0x138
> container/list.(*List).PushBack()
> /usr/local/go/src/container/list/list.go:155 +0x11b
> github.com/apolloconfig/agollo/v4/storage.(*Cache).AddChangeListener()
> /data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/fananchong/agollo/v4@v4.1.2-0.20220504001018-a7beadea6359/storage/repository.go:377 +0x1d
> github.com/apolloconfig/agollo/v4.(*internalClient).AddChangeListener()
> /data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/fananchong/agollo/v4@v4.1.2-0.20220504001018-a7beadea6359/client.go:246 +0x54
> haidao/backend/core/configv2.(*cfgApollo).startCommonNamespace()
> /tmp/gocbuildbe8172cde925haidao_backend_cross_season/core/configv2/cfg_apollo.go:104 +0x2b8
> haidao/backend/core/configv2.(*cfgApollo).Init()
> /tmp/gocbuildbe8172cde925haidao_backend_cross_season/core/configv2/cfg_apollo.go:59 +0x828
> haidao/backend/core/configv2.Init()
> /tmp/gocbuildbe8172cde925haidao_backend_cross_season/core/configv2/cfg.go:44 +0x4f7
> haidao/backend/core/app.Run.func2()
> /tmp/gocbuildbe8172cde925haidao_backend_cross_season/core/app/app.go:128 +0x8ec
> github.com/urfave/cli/v2.(*App).RunContext()
> /data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/urfave/cli/v2@v2.5.1/app.go:322 +0xdd5
> github.com/urfave/cli/v2.(*App).Run()
> /data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/urfave/cli/v2@v2.5.1/app.go:224 +0x16a8
> haidao/backend/core/app.Run()
> /tmp/gocbuildbe8172cde925haidao_backend_cross_season/core/app/app.go:167 +0x162c
> main.main()
> /tmp/gocbuildbe8172cde925haidao_backend_cross_season/cross_season/main.go:30 +0x50
> 
> Goroutine 52 (running) created at:
> github.com/apolloconfig/agollo/v4.StartWithConfig()
> /data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/fananchong/agollo/v4@v4.1.2-0.20220504001018-a7beadea6359/client.go:137 +0x696
> haidao/backend/core/configv2.(*cfgApollo).startCommonNamespace()
> /tmp/gocbuildbe8172cde925haidao_backend_cross_season/core/configv2/cfg_apollo.go:88 +0x14b
> haidao/backend/core/configv2.(*cfgApollo).Init()
> /tmp/gocbuildbe8172cde925haidao_backend_cross_season/core/configv2/cfg_apollo.go:59 +0x828
> haidao/backend/core/configv2.Init()
> /tmp/gocbuildbe8172cde925haidao_backend_cross_season/core/configv2/cfg.go:44 +0x4f7
> haidao/backend/core/app.Run.func2()
> /tmp/gocbuildbe8172cde925haidao_backend_cross_season/core/app/app.go:128 +0x8ec
> github.com/urfave/cli/v2.(*App).RunContext()
> /data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/urfave/cli/v2@v2.5.1/app.go:322 +0xdd5
> github.com/urfave/cli/v2.(*App).Run()
> /data/workspace/uto_deploy_build_weeklineWithUGS/weeklineWithUGS/backend/pkg/mod/github.com/urfave/cli/v2@v2.5.1/app.go:224 +0x16a8
> haidao/backend/core/app.Run()
> /tmp/gocbuildbe8172cde925haidao_backend_cross_season/core/app/app.go:167 +0x162c
> main.main()
> /tmp/gocbuildbe8172cde925haidao_backend_cross_season/cross_season/main.go:30 +0x50
> ```
> 
> storage.(*Cache).AddChangeListener() 与 storage.(*Cache).pushChange() 间有 race

感谢提供这么重要的线索，我觉得 data race 是可以消除的。
* 将其中所有不涉及修改的读取 c.changeListeners 的操作，映射到 ```func (c *Cache) GetChangeListeners() *list.List``` 并在其中加上读锁。
* 在 ```AddChangeListener ``` 和 ``` RemoveChangeListener ``` 加上写锁。

Comment (zouyx) [reply client.go:138]: #231

**Pull request closed without merging**